src/openvic-simulation/diplomacy/CountryRelation.hpp

```diff
@@ -46,50 +48,180 @@ namespace OpenVic {
+			Neutral,
+			Cordial,
+			Friendly,
+			Sphere
+		};
+
 	private:
-		// TODO: reference of manager responsible for storing CountryInstances
-		ordered_map<CountryRelationPair, country_relation_value_t> country_relations;
-
-	public:
-		CountryRelationManager(/* TODO: Country Instance Manager Reference */);
-
-		bool add_country(CountryRelationInstanceProxy country);
-		bool remove_country(CountryRelationInstanceProxy country);
-
-		country_relation_value_t
-		get_country_relation(CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient) const;
-		country_relation_value_t*
-		get_country_relation_ptr(CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient);
-		bool set_country_relation(
-			CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient, country_relation_value_t value
-		);
+#define RELATION_PAIR_MAP(PAIR_TYPE, VALUE_TYPE, NAME, FUNC_NAME, RECEPIENT_CONST) \
+	deque_ordered_map<PAIR_TYPE, VALUE_TYPE> NAME; \
+\
+public: \
+	VALUE_TYPE get_##FUNC_NAME(CountryInstance const* country, CountryInstance const* recepient) const; \
+	VALUE_TYPE& assign_or_get_##FUNC_NAME( \
+		CountryInstance* country, CountryInstance RECEPIENT_CONST* recepient, VALUE_TYPE default_value = {} \
+	); \
+	bool set_##FUNC_NAME(CountryInstance* country, CountryInstance RECEPIENT_CONST* recepient, VALUE_TYPE value); \
+	decltype(NAME)::values_container_type const& get_##FUNC_NAME##_values() const; \
+\
+private:
+
+		RELATION_PAIR_MAP(UnorderedCountryInstancePair, relation_value_type, relations, country_relation, );
+		RELATION_PAIR_MAP(UnorderedCountryInstancePair, bool, alliances, country_alliance, );
+		RELATION_PAIR_MAP(UnorderedCountryInstancePair, bool, at_war, country_at_war, );
+
+		RELATION_PAIR_MAP(CountryInstancePair, bool, military_access, country_military_access, const);
+		RELATION_PAIR_MAP(CountryInstancePair, bool, war_subsidies, country_war_subsidies, const);
+		RELATION_PAIR_MAP(CountryInstancePair, bool, command_units, country_command_units, const);
+		RELATION_PAIR_MAP(CountryInstancePair, bool, vision, country_vision, const);
+		RELATION_PAIR_MAP(CountryInstancePair, OpinionType, opinions, country_opinion, const);
+		RELATION_PAIR_MAP(CountryInstancePair, influence_value_type, influence, country_influence, const);
+		RELATION_PAIR_MAP(CountryInstancePair, bool, discredits, country_discredit, const);
+		RELATION_PAIR_MAP(CountryInstancePair, bool, embassy_bans, country_embassy_ban, const);
+
+#undef RELATION_PAIR_MAP
 	};
 }
+
+constexpr OpenVic::CountryRelationManager::OpinionType& operator++(OpenVic::CountryRelationManager::OpinionType& type) {
+	using underlying_type = std::underlying_type_t<OpenVic::CountryRelationManager::OpinionType>;
+	type = static_cast<OpenVic::CountryRelationManager::OpinionType>(static_cast<underlying_type>(type) + 1);
+	type = static_cast<underlying_type>(type) >
+			static_cast<underlying_type>(OpenVic::CountryRelationManager::OpinionType::Sphere) //
```

Comment:
```suggestion
			static_cast<underlying_type>(OpenVic::CountryRelationManager::OpinionType::Sphere)
```
Also keep on same line as `type = static_cast<underlying_type>(type) >`
Now it's just floating there weirdly.